contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.20;
+
+import "forge-std/console2.sol";
+import "../libraries/SwapIntention.sol";
+import {IQuoter} from "../interfaces/IQuoter.sol";
+import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {SafeCast} from "@uniswap/v4-core/contracts/libraries/SafeCast.sol";
+import {Pool} from "@uniswap/v4-core/contracts/libraries/Pool.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {PoolManager} from "@uniswap/v4-core/contracts/PoolManager.sol";
+
+contract Quoter is IQuoter {
+    using PoolIdLibrary for PoolKey;
+    using SafeCast for *;
+    using Hooks for IHooks;
+
+    // v4 Singleton contract
+    IPoolManager poolManager;
```

Comment:
since it is defined in constructor and never changes, you can specify it as `immutable` which has big gas benefits as it's written into the contract code rather than sloaded every time its read

Reply:
ik gas not super important here since its a quoter, but does help us squeeze more quotes into one multicall